syntax = "proto3";

import "roi.proto";
import "roivolume.proto";
import "dcmpix.proto";
import "vrcontroller.proto";
import "viewercontroller.proto";
import "browsercontroller.proto";
import "dicomimage.proto";
import "dicomseries.proto";
import "dicomstudy.proto";
import "utilities.proto";
import "types.proto";

package osirixgrpc;

service OsiriXService {

// OsiriX methods
rpc OsirixCurrentBrowser (Empty) returns (OsirixCurrentBrowserResponse) {}
rpc OsirixFrontmostViewer (Empty) returns (OsirixFrontmostViewerResponse) {}
rpc OsirixDisplayed2DViewers (Empty) returns (OsirixDisplayed2DViewersResponse) {}
rpc OsirixFrontmostVRController (Empty) returns (OsirixFrontmostVRControllerResponse) {}
rpc OsirixDisplayedVRControllers (Empty) returns (OsirixDisplayedVRControllersResponse) {}
rpc OsirixVersion (Empty) returns (OsirixVersionResponse) {}

// Cache methods - debug only
rpc OsirixCacheUids (Empty) returns (OsirixCacheUidsResponse) {}
rpc OsirixCacheObjectForUid (OsirixCacheObjectForUidRequest) returns (OsirixCacheObjectForUidResponse) {}

// ROI methods
rpc ROIDelete (ROI) returns (Response) {}
rpc ROIFlipHorizontally (ROI) returns (Response) {}
rpc ROIFlipVertically (ROI) returns (Response) {}
rpc ROIArea (ROI) returns (ROIAreaResponse) {}
rpc ROICentroid (ROI) returns (ROICentroidResponse) {}
rpc ROIRotate (ROIRotateRequest) returns (Response) {}
rpc ROIMove (ROIMoveRequest) returns (Response) {}
rpc ROIPix (ROI) returns (ROIPixResponse) {}
rpc ROIName (ROI) returns (ROINameResponse) {}
rpc ROIIType (ROI) returns (ROIITypeResponse) {}
rpc ROISetName (ROISetNameRequest) returns (Response) {}
rpc ROIColor (ROI) returns (ROIColorResponse) {}
rpc ROISetColor (ROISetColorRequest) returns (Response) {}
rpc ROIOpacity (ROI) returns (ROIOpacityResponse) {}
rpc ROISetOpacity (ROISetOpacityRequest) returns (Response) {}
rpc ROIThickness (ROI) returns (ROIThicknessResponse) {}
rpc ROISetThickness (ROISetThicknessRequest) returns (Response) {}
rpc ROIPoints (ROI) returns (ROIPointsResponse) {}
rpc ROISetPoints (ROISetPointsRequest) returns (Response) {}

// ROIVolume methods
rpc ROIVolumeTexture (ROIVolume) returns (ROIVolumeTextureResponse) {}
rpc ROIVolumeSetTexture (ROIVolumeSetTextureRequest) returns (Response) {}
rpc ROIVolumeVolume (ROIVolume) returns (ROIVolumeVolumeResponse) {}
rpc ROIVolumeColor(ROIVolume) returns (ROIVolumeColorResponse) {}
rpc ROIVolumeSetColor(ROIVolumeSetColorRequest) returns (Response) {}
rpc ROIVolumeOpacity(ROIVolume) returns (ROIVolumeOpacityResponse) {}
rpc ROIVolumeSetOpacity(ROIVolumeSetOpacityRequest) returns (Response) {}
rpc ROIVolumeFactor(ROIVolume) returns (ROIVolumeFactorResponse) {}
rpc ROIVolumeSetFactor(ROIVolumeSetFactorRequest) returns (Response) {}
rpc ROIVolumeName(ROIVolume) returns (ROIVolumeNameResponse) {}
rpc ROIVolumeVisible (ROIVolume) returns (ROIVolumeVisibleResponse) {}

// DCMPix methods
rpc DCMPixConvertToRGB (DCMPixConvertToRGBRequest) returns (Response) {}
rpc DCMPixConvertToBW (DCMPixConvertToBWRequest) returns (Response) {}
rpc DCMPixIsRGB (DCMPix) returns (DCMPixIsRGBResponse) {}
rpc DCMPixComputeROI (DCMPixComputeROIRequest) returns (DCMPixComputeROIResponse) {}
rpc DCMPixROIValues (DCMPixROIValuesRequest) returns (DCMPixROIValuesResponse) {}
rpc DCMPixShape (DCMPix) returns (DCMPixShapeResponse) {}
rpc DCMPixSpacing (DCMPix) returns (DCMPixSpacingResponse) {}
rpc DCMPixOrigin (DCMPix) returns (DCMPixOriginResponse) {}
rpc DCMPixOrientation (DCMPix) returns (DCMPixOrientationResponse) {}
rpc DCMPixSliceLocation (DCMPix) returns (DCMPixSliceLocationResponse) {}
rpc DCMPixSourceFile (DCMPix) returns (DCMPixSourceFileResponse) {}
rpc DCMPixImage (DCMPix) returns (DCMPixImageResponse) {}
rpc DCMPixSetImage (DCMPixSetImageRequest) returns (Response) {}
rpc DCMPixGetMapFromROI (DCMPixGetMapFromROIRequest) returns (DCMPixGetMapFromROIResponse) {}
rpc DCMPixDicomImage (DCMPix) returns (DCMPixDicomImageResponse) {}
rpc DCMPixDicomSeries (DCMPix) returns (DCMPixDicomSeriesResponse) {}
rpc DCMPixDicomStudy (DCMPix) returns (DCMPixDicomStudyResponse) {}

// VRController methods
rpc VRControllerViewer2D (VRController) returns (VRControllerViewer2DResponse) {}
rpc VRControllerBlendingController (VRController) returns (VRControllerBlendingControllerResponse) {}
rpc VRControllerStyle (VRController) returns (VRControllerStyleResponse) {}
rpc VRControllerTitle (VRController) returns (VRControllerTitleResponse) {}
rpc VRControllerROIVolumes (VRController) returns (VRControllerROIVolumesResponse) {}
rpc VRControllerRenderingMode (VRController) returns (VRControllerRenderingModeResponse) {}
rpc VRControllerSetRenderingMode (VRControllerSetRenderingModeRequest) returns (Response) {}
rpc VRControllerWLWW (VRController) returns (VRControllerWLWWResponse) {}
rpc VRControllerSetWLWW (VRControllerSetWLWWRequest) returns (Response) {}
rpc VRControllerHideROIVolume (VRControllerHideROIVolumeRequest) returns (Response) {}
rpc VRControllerDisplayROIVolume (VRControllerDisplayROIVolumeRequest) returns (Response) {}
rpc VRControllerNeedsDisplayUpdate (VRController) returns (Response) {}

// ViewerController methods
rpc ViewerControllerCloseViewer (ViewerController) returns (Response) {}
rpc ViewerControllerPixList (ViewerControllerPixListRequest) returns (ViewerControllerPixListResponse) {}
rpc ViewerControllerNeedsDisplayUpdate (ViewerController) returns (Response) {}
rpc ViewerControllerROIList (ViewerControllerROIListRequest) returns (ViewerControllerROIListResponse) {}
rpc ViewerControllerNewROI (ViewerControllerNewROIRequest) returns (ViewerControllerNewROIResponse) {}
rpc ViewerControllerCurDCM (ViewerController) returns (ViewerControllerCurDCMResponse) {}
rpc ViewerControllerROIsWithName (ViewerControllerROIsWithNameRequest) returns (ViewerControllerROIsWithNameResponse) {}
rpc ViewerControllerSelectedROIs (ViewerController) returns (ViewerControllerSelectedROIsResponse) {}
rpc ViewerControllerIsDataVolumic (ViewerControllerIsDataVolumicRequest) returns (ViewerControllerIsDataVolumicResponse) {}
rpc ViewerControllerCopyViewerWindow (ViewerController) returns (ViewerControllerCopyViewerWindowResponse) {}
rpc ViewerControllerResampleViewerController (ViewerControllerResampleViewerControllerRequest) returns (ViewerControllerResampleViewerControllerResponse) {}
rpc ViewerControllerBlendingController (ViewerController) returns (ViewerControllerBlendingControllerResponse) {}
rpc ViewerControllerVRControllers (ViewerController) returns (ViewerControllerVRControllersResponse) {}
rpc ViewerControllerTitle (ViewerController) returns (ViewerControllerTitleResponse) {}
rpc ViewerControllerModality (ViewerController) returns (ViewerControllerModalityResponse) {}
rpc ViewerControllerMovieIdx (ViewerController) returns (ViewerControllerMovieIdxResponse) {}
rpc ViewerControllerSetMovieIdx (ViewerControllerSetMovieIdxRequest) returns (Response) {}
rpc ViewerControllerMaxMovieIdx (ViewerController) returns (ViewerControllerMaxMovieIdxResponse) {}
rpc ViewerControllerIdx (ViewerController) returns (ViewerControllerIdxResponse) {}
rpc ViewerControllerSetIdx (ViewerControllerSetIdxRequest) returns (Response) {}
rpc ViewerControllerDisplayedIdx (ViewerController) returns (ViewerControllerDisplayedIdxResponse) {}
rpc ViewerControllerSetDisplayedIdx (ViewerControllerSetDisplayedIdxRequest) returns (Response) {}
rpc ViewerControllerFlippedData (ViewerController) returns (ViewerControllerFlippedDataResponse) {}
rpc ViewerControllerWLWW (ViewerController) returns (ViewerControllerWLWWResponse) {}
rpc ViewerControllerSetWLWW (ViewerControllerSetWLWWRequest) returns (Response) {}
rpc ViewerControllerOpenVRViewerForMode (ViewerControllerOpenVRViewerForModeRequest) returns (ViewerControllerOpenVRViewerForModeResponse) {}
rpc ViewerControllerFuseWithViewer (ViewerControllerFuseWithViewerRequest) returns (Response) {}

// BrowserController methods
rpc BrowserControllerDatabaseSelection (BrowserController) returns (BrowserControllerDatabaseSelectionResponse) {}
rpc BrowserControllerCopyFilesIfNeeded (BrowserControllerCopyFilesIfNeededRequest) returns (Response) {}
rpc BrowserControllerOpenViewerFromImages (BrowserControllerOpenViewerFromImagesRequest) returns (BrowserControllerOpenViewerFromImagesResponse) {}
rpc BrowserControllerDatabaseStudies (BrowserController) returns (BrowserControllerDatabaseStudiesResponse) {}

// DicomImage methods
rpc DicomImageWidth (DicomImage) returns (DicomImageWidthResponse) {}
rpc DicomImageHeight (DicomImage) returns (DicomImageHeightResponse) {}
rpc DicomImageSOPInstanceUID (DicomImage) returns (DicomImageSOPInstanceUIDResponse) {}
rpc DicomImageCompletePath (DicomImage) returns (DicomImageCompletePathResponse) {}
rpc DicomImageDate (DicomImage) returns (DicomImageDateResponse) {}
rpc DicomImageNumberOfFrames (DicomImage) returns (DicomImageNumberOfFramesResponse) {}
rpc DicomImageModality (DicomImage) returns (DicomImageModalityResponse) {}
rpc DicomImageSeries (DicomImage) returns (DicomImageSeriesResponse) {}
rpc DicomImageSliceLocation (DicomImage) returns (DicomImageSliceLocationResponse) {}
rpc DicomImageInstanceNumber (DicomImage) returns (DicomImageInstanceNumberResponse) {}

// DicomSeries methods
rpc DicomSeriesPaths (DicomSeries) returns (DicomSeriesPathsResponse) {}
rpc DicomSeriesPreviousSeries (DicomSeries) returns (DicomSeriesPreviousSeriesResponse) {}
rpc DicomSeriesNextSeries (DicomSeries) returns (DicomSeriesNextSeriesResponse) {}
rpc DicomSeriesSortedImages (DicomSeries) returns (DicomSeriesSortedImagesResponse) {}
rpc DicomSeriesStudy (DicomSeries) returns (DicomSeriesStudyResponse) {}
rpc DicomSeriesImages (DicomSeries) returns (DicomSeriesImagesResponse) {}
rpc DicomSeriesSeriesInstanceUID (DicomSeries) returns (DicomSeriesSeriesInstanceUIDResponse) {}
rpc DicomSeriesSeriesSOPClassUID (DicomSeries) returns (DicomSeriesSeriesSOPClassUIDResponse) {}
rpc DicomSeriesSeriesDescription (DicomSeries) returns (DicomSeriesSeriesDescriptionResponse) {}
rpc DicomSeriesModality (DicomSeries) returns (DicomSeriesModalityResponse) {}
rpc DicomSeriesName (DicomSeries) returns (DicomSeriesNameResponse) {}
rpc DicomSeriesDate (DicomSeries) returns (DicomSeriesDateResponse) {}
rpc DicomSeriesNumberOfImages (DicomSeries) returns (DicomSeriesNumberOfImagesResponse) {}

// DicomStudy methods
rpc DicomStudyPaths (DicomStudy) returns (DicomStudyPathsResponse) {}
rpc DicomStudyImages (DicomStudy) returns (DicomStudyImagesResponse) {}
rpc DicomStudyModalities (DicomStudy) returns (DicomStudyModalitiesResponse) {}
rpc DicomStudyNoFiles (DicomStudy) returns (DicomStudyNoFilesResponse) {}
rpc DicomStudyRawNoFiles (DicomStudy) returns (DicomStudyRawNoFilesResponse) {}
rpc DicomStudyNumberOfImages (DicomStudy) returns (DicomStudyNumberOfImagesResponse) {}
rpc DicomStudySeries (DicomStudy) returns (DicomStudySeriesResponse) {}
rpc DicomStudyName (DicomStudy) returns (DicomStudyNameResponse) {}
rpc DicomStudyDate (DicomStudy) returns (DicomStudyDateResponse) {}
rpc DicomStudyDateAdded (DicomStudy) returns (DicomStudyDateAddedResponse) {}
rpc DicomStudyDateOfBirth (DicomStudy) returns (DicomStudyDateOfBirthResponse) {}
rpc DicomStudyInstitutionName (DicomStudy) returns (DicomStudyInstitutionNameResponse) {}
rpc DicomStudyModality (DicomStudy) returns (DicomStudyModalityResponse) {}
rpc DicomStudyPatientID (DicomStudy) returns (DicomStudyPatientIDResponse) {}
rpc DicomStudyPatientUID (DicomStudy) returns (DicomStudyPatientUIDResponse) {}
rpc DicomStudyPatientSex (DicomStudy) returns (DicomStudyPatientSexResponse) {}
rpc DicomStudyPerformingPhysician (DicomStudy) returns (DicomStudyPerformingPhysicianResponse) {}
rpc DicomStudyReferringPhysician (DicomStudy) returns (DicomStudyReferringPhysicianResponse) {}
rpc DicomStudyStudyInstanceUID (DicomStudy) returns (DicomStudyStudyInstanceUIDResponse) {}
rpc DicomStudyStudyName (DicomStudy) returns (DicomStudyStudyNameResponse) {}
}

// Response messages
message OsirixCurrentBrowserResponse{
	Status status = 1;
	BrowserController browser_controller = 2;
}

message OsirixFrontmostViewerResponse{
	Status status = 1;
	ViewerController viewer_controller = 2;
}

message OsirixDisplayed2DViewersResponse{
	Status status = 1;
	repeated ViewerController viewer_controllers = 2;
}

message OsirixFrontmostVRControllerResponse{
	Status status = 1;
	VRController vr_controller = 2;
}
 
message OsirixDisplayedVRControllersResponse{
	Status status = 1;
	repeated VRController vr_controllers = 2;
}

message OsirixVersionResponse{
	Status status = 1;
	string version = 2;
	string bundle_name = 3;
}


// Cache requests / responses
message OsirixCacheUidsResponse {
	Status status = 1;
	repeated string uids = 2;
}

message OsirixCacheObjectForUidResponse {
	Status status = 1;
	string object_type = 2;
}

message OsirixCacheObjectForUidRequest {
	string uid = 1;
}
